datafusion/physical-expr/src/utils/guarantee.rs

```diff
@@ -645,9 +680,107 @@ mod test {
+            vec![not_in_guarantee("b", [1, 2, 3])],
+        );
+    }
+
+    #[test]
+    fn test_inlist_conjunction() {
+        // b IN (1, 2, 3) AND b IN (2, 3, 4)
+        test_analyze(
+            col("b")
+                .in_list(vec![lit(1), lit(2), lit(3)], false)
+                .and(col("b").in_list(vec![lit(2), lit(3), lit(4)], false)),
+            vec![in_guarantee("b", [2, 3])],
+        );
+        // b NOT IN (1, 2, 3) AND b IN (2, 3, 4)
+        test_analyze(
+            col("b")
+                .in_list(vec![lit(1), lit(2), lit(3)], true)
+                .and(col("b").in_list(vec![lit(2), lit(3), lit(4)], false)),
+            vec![
+                not_in_guarantee("b", [1, 2, 3]),
+                in_guarantee("b", [2, 3, 4]),
+            ],
+        );
+        // b NOT IN (1, 2, 3) AND b NOT IN (2, 3, 4)
+        test_analyze(
+            col("b")
+                .in_list(vec![lit(1), lit(2), lit(3)], true)
+                .and(col("b").in_list(vec![lit(2), lit(3), lit(4)], true)),
+            vec![not_in_guarantee("b", [1, 2, 3, 4])],
+        );
+        // b IN (1, 2, 3) AND b = 4
+        test_analyze(
+            col("b")
+                .in_list(vec![lit(1), lit(2), lit(3)], false)
+                .and(col("b").eq(lit(4))),
+            vec![],
+        );
+        // b IN (1, 2, 3) AND b = 2
+        test_analyze(
+            col("b")
+                .in_list(vec![lit(1), lit(2), lit(3)], false)
+                .and(col("b").eq(lit(2))),
+            vec![in_guarantee("b", [2])],
+        );
+        // b IN (1, 2, 3) AND b != 2
+        test_analyze(
+            col("b")
+                .in_list(vec![lit(1), lit(2), lit(3)], false)
+                .and(col("b").not_eq(lit(2))),
+            vec![in_guarantee("b", [1, 2, 3]), not_in_guarantee("b", [2])],
+        );
+        // b NOT IN (1, 2, 3) AND b != 4
+        test_analyze(
+            col("b")
+                .in_list(vec![lit(1), lit(2), lit(3)], true)
+                .and(col("b").not_eq(lit(4))),
+            vec![not_in_guarantee("b", [1, 2, 3, 4])],
+        );
+        // b NOT IN (1, 2, 3) AND b != 2
+        test_analyze(
+            col("b")
+                .in_list(vec![lit(1), lit(2), lit(3)], true)
+                .and(col("b").not_eq(lit(2))),
+            vec![not_in_guarantee("b", [1, 2, 3])],
+        );
+    }
+
+    #[test]
+    fn test_inlist_with_disjunction() {
+        // b IN (1, 2, 3) AND (b = 3 OR b = 4)
+        test_analyze(
+            col("b")
+                .in_list(vec![lit(1), lit(2), lit(3)], false)
+                .and(col("b").eq(lit(3)).or(col("b").eq(lit(4)))),
+            vec![in_guarantee("b", [3])],
+        );
+        // b IN (1, 2, 3) AND (b = 4 OR b = 5)
+        test_analyze(
+            col("b")
+                .in_list(vec![lit(1), lit(2), lit(3)], false)
+                .and(col("b").eq(lit(4)).or(col("b").eq(lit(5)))),
+            vec![],
```

Comment:
I theory this could be `in_guarantee(1,2,3)` right? As it can only be true when b is 1,2,3.
However in this case this expression can never be true so maybe it doesn't really matter 🤔

Reply:
This would be invalid since intersection between [1,2,3] and [4,5] is empty.
```rust
let intersection = new_values
    .into_iter()
    .filter(|new_value| existing.literals.contains(*new_value))
    .collect::<Vec<_>>();
// for an In guarantee, if the intersection is not empty,  we can extend the guarantee
// e.g. `a IN (1,2,3) AND a IN (2,3,4)` is `a IN (2,3)`
// otherwise, we invalidate the guarantee
// e.g. `a IN (1,2,3) AND a IN (4,5,6)` is `a IN ()`, which is invalid
if !intersection.is_empty() {
    existing.literals = intersection.into_iter().cloned().collect();
} else {
    // at least one was not, so invalidate the guarantee
    *entry = None;
}
```

BTW,  I left a comment in #8437. Please cc when you are free.

Reply:
> This would be invalid since intersection between [1,2,3] and [4,5] is empty.

Right -- I wonder if in general for cases where the intersection is empty, can we infer that the expression can not be true ever 🤔  (there is no way to represent this today in the guarantee framework, maybe something like 


```
enum Guarantee {
  /// This predicate can not be true
  CanNotBeTrue,
  /// This predicate can only be true if all the `LiteralGurantee`s are valie
  List(Vec<LiteralGurantee>)
}

And then return `Gurantee::CanNotBeTrue` if we discover that the expression can not be true 🤔 


> BTW, I left a comment in https://github.com/apache/arrow-datafusion/pull/8437. Please cc when you are free.

I am not quite sure what you are referring to. https://github.com/apache/arrow-datafusion/pull/8437 doesn't seem to have had any recent activity that I can see.

Reply:
Maybe we can enhance `eliminate_filter` optimizer rule to eliminate the filters(in where clause) that always not be true and replace by a `EmptyRelation`.
like we do for `where false` in 
https://github.com/apache/arrow-datafusion/blob/8284371cb5dbeb5d0b1d50c420affb9be86b1599/datafusion/optimizer/src/eliminate_filter.rs#L55-L62

Reply:
I filed this idea in https://github.com/apache/arrow-datafusion/issues/8687